### PR TITLE
Add access group template support

### PR DESCRIPTION
**Pull request by hariarla**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...


```
Acceptance Tests:
image (18)
image
```
=== RUN   TestAccIBMIAMAccessGroupTemplateAssignmentBasic
--- PASS: TestAccIBMIAMAccessGroupTemplateAssignmentBasic (180.99s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/iamaccessgroup	183.432s
```
image
image
Create base template, a new version underneath it.
```
resource "ibm_iam_access_group_template" "template" {
    account_id          = "dfc6bf8c040145f8b4dd20973c0bdbfd"
    committed           = false
    created_at          = "2023-08-12T10:47:02.000Z"
    created_by_id       = "IBMid-668000IN8T"
    description         = "Testing"
    href                = "https://iam.cloud.ibm.com/v1/group_templates/AccessGroupTemplateId-e2fa9f02-a95e-4450-9fb3-db102cccf0f4/versions/1"
    id                  = "AccessGroupTemplateId-e2fa9f02-a95e-4450-9fb3-db102cccf0f4/1"
    last_modified_at    = "2023-08-12T10:47:02.000Z"
    last_modified_by_id = "IBMid-668000IN8T"
    name                = "TestAGTemplatesTerraform345678"
    template_id         = "AccessGroupTemplateId-e2fa9f02-a95e-4450-9fb3-db102cccf0f4"
    version             = "1"

    group {
        name = "IAM Admin Group Test 3456789"

        action_controls {
            access {
                add = false
            }
        }

        assertions {
            action_controls {
                add    = false
                remove = false
                update = false
            }
        }

        members {
            services = []
            users    = []

            action_controls {
                add    = false
                remove = false
            }
        }
    }
}

# ibm_iam_access_group_template_version.template:
resource "ibm_iam_access_group_template_version" "template" {
    account_id          = "dfc6bf8c040145f8b4dd20973c0bdbfd"
    committed           = false
    created_at          = "2023-08-12T10:47:04.000Z"
    created_by_id       = "IBMid-668000IN8T"
    description         = "Testing"
    href                = "https://iam.cloud.ibm.com/v1/group_templates/AccessGroupTemplateId-e2fa9f02-a95e-4450-9fb3-db102cccf0f4/versions/2"
    id                  = "AccessGroupTemplateId-e2fa9f02-a95e-4450-9fb3-db102cccf0f4/2"
    last_modified_at    = "2023-08-12T10:47:04.000Z"
    last_modified_by_id = "IBMid-668000IN8T"
    name                = "TestAGTemplatesTerraform345678910"
    template_id         = "AccessGroupTemplateId-e2fa9f02-a95e-4450-9fb3-db102cccf0f4"
    version             = "2"

    group {
        name = "IAM Admin Group Test 34567891011"

        action_controls {
            access {
                add = false
            }
        }

        assertions {
            action_controls {
                add    = false
                remove = false
                update = false
            }
        }

        members {
            services = []
            users    = []

            action_controls {
                add    = false
                remove = false
            }
        }
    }
}
```
List template versions
```
terraform show
# data.ibm_iam_access_group_template_version.template:
data "ibm_iam_access_group_template_version" "template" {
    account_id                   = "dfc6bf8c040145f8b4dd20973c0bdbfd"
    committed                    = false
    created_at                   = "2023-08-13T06:33:42.000Z"
    created_by_id                = "IBMid-668000IN8T"
    description                  = "Testing4"
    group                        = [
        {
            action_controls = [
                {
                    access = [
                        {
                            add = false
                        },
                    ]
                },
            ]
            assertions      = [
                {
                    action_controls = [
                        {
                            add    = false
                            remove = true
                            update = true
                        },
                    ]
                    rules           = []
                },
            ]
            description     = ""
            members         = [
                {
                    action_controls = [
                        {
                            add    = false
                            remove = false
                        },
                    ]
                    services        = []
                    users           = []
                },
            ]
            name            = "IAM Admin Group Test 34567891011"
        },
    ]
    href                         = "https://iam.cloud.ibm.com/v1/group_templates/AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921/versions/2"
    id                           = "AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921"
    last_modified_at             = "2023-08-13T06:59:28.000Z"
    last_modified_by_id          = "IBMid-668000IN8T"
    name                         = "TestAGTemplatesTerraform345678"
    policy_template_references   = []
    template_id                  = "AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921"
    template_version_response_id = "AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921"
    verbose                      = true
    version                      = "2"
}
```

Create Assignment:
```
terraform show
# ibm_iam_access_group_template_assignment.assginment:
resource "ibm_iam_access_group_template_assignment" "assginment" {
    account_id          = "dfc6bf8c040145f8b4dd20973c0bdbfd"
    created_at          = "2023-08-13T17:46:53.000Z"
    created_by_id       = "IBMid-668000IN8T"
    etag                = "2-c48fe600d397566bde8a44323fa4adeb"
    href                = "https://iam.cloud.ibm.com/v1/group_assignments/AccessGroupAssignmentId-AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921-Target-b5ba2be0a9c44233a2f8b600992a5523"
    id                  = "AccessGroupAssignmentId-AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921-Target-b5ba2be0a9c44233a2f8b600992a5523"
    last_modified_at    = "2023-08-13T17:47:42.000Z"
    last_modified_by_id = "IBMid-668000IN8T"
    operation           = "assign"
    status              = "succeeded"
    target              = "b5ba2be0a9c44233a2f8b600992a5523"
    target_type         = "AccountGroup"
    template_id         = "AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921"
    template_version    = "2"
}
```
Update Assignment:
```
terraform show
# ibm_iam_access_group_template_assignment.assginment:
resource "ibm_iam_access_group_template_assignment" "assginment" {
    account_id          = "dfc6bf8c040145f8b4dd20973c0bdbfd"
    created_at          = "2023-08-13T17:46:53.000Z"
    created_by_id       = "IBMid-668000IN8T"
    etag                = "4-cc34f7261ab8db99896c8be8d6a627ea"
    href                = "https://iam.cloud.ibm.com/v1/group_assignments/AccessGroupAssignmentId-AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921-Target-b5ba2be0a9c44233a2f8b600992a5523"
    id                  = "AccessGroupAssignmentId-AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921-Target-b5ba2be0a9c44233a2f8b600992a5523"
    last_modified_at    = "2023-08-13T17:58:48.000Z"
    last_modified_by_id = "IBMid-668000IN8T"
    operation           = "update"
    status              = "succeeded"
    target              = "b5ba2be0a9c44233a2f8b600992a5523"
    target_type         = "AccountGroup"
    template_id         = "AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921"
    template_version    = "1"
}
```

List Assignment:
```
terraform show
# data.ibm_iam_access_group_template_assignment.template:
data "ibm_iam_access_group_template_assignment" "template" {
    account_id  = "dfc6bf8c040145f8b4dd20973c0bdbfd"
    assignments = [
        {
            account_id          = "dfc6bf8c040145f8b4dd20973c0bdbfd"
            created_at          = "2023-08-13T18:10:43.000Z"
            created_by_id       = "IBMid-668000IN8T"
            href                = "https://iam.cloud.ibm.com/v1/group_assignments/AccessGroupAssignmentId-AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921-Target-b5ba2be0a9c44233a2f8b600992a5523"
            id                  = "AccessGroupAssignmentId-AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921-Target-b5ba2be0a9c44233a2f8b600992a5523"
            last_modified_at    = "2023-08-13T18:11:00.000Z"
            last_modified_by_id = "IBMid-668000IN8T"
            operation           = "assign"
            status              = "succeeded"
            target              = "b5ba2be0a9c44233a2f8b600992a5523"
            target_type         = "AccountGroup"
            template_id         = "AccessGroupTemplateId-dd8dfdc0-fa8e-4de7-b6d3-19f2a1a21921"
            template_version    = "1"
        },
    ]
    first       = [
        {
            href = "https://iam.cloud.ibm.com/v1/group_assignments?limit=50&account_id=dfc6bf8c040145f8b4dd20973c0bdbfd"
        },
    ]
    id          = "2023-08-13 18:12:03.34375 +0000 UTC"
    last        = [
        {
            href = "https://iam.cloud.ibm.com/v1/group_assignments?offset=0&limit=50&account_id=dfc6bf8c040145f8b4dd20973c0bdbfd"
        },
    ]
    limit       = 50
    offset      = 0
    total_count = 1
}
```

### CODE REVIEW

File: ibm/service/iamaccessgroup/resource_ibm_iam_access_group_template_assignment.go [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+	}
+	if err = d.Set("etag", response.Headers.Get("Etag")); err != nil {
+		return diag.FromErr(fmt.Errorf("Error setting etag: %s", err))
+	}
+
+	return nil
+}
+
+func resourceIBMIamAccessGroupTemplateAssignmentUpdate(context context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	fmt.Println("==========update===================")
```
Comment (hkantare) [inline ibm/service/iamaccessgroup/resource_ibm_iam_access_group_template_assignment.go:269]: remove this  kind of fmt stmts

File: ibm/service/iamaccessgroup/resource_ibm_iam_access_group_template_assignment.go [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+		log.Printf("[DEBUG] GetAssignmentWithContext failed %s\n%s", err, response)
+		return diag.FromErr(fmt.Errorf("GetAssignmentWithContext failed %s\n%s", err, response))
+	}
+
+	updateAssignmentOptions.SetIfMatch(response.Headers.Get("ETag"))
+
+	hasChange := false
+
+	if d.HasChange("template_version") {
+		fmt.Println("==========update1===================")
```
Comment (hkantare) [inline ibm/service/iamaccessgroup/resource_ibm_iam_access_group_template_assignment.go:299]: Only "template_version" can be updated . Remaining arguments in schema should we add ForceNew on them ?

File: website/docs/d/iam_access_group_template.html.markdown [hunk 3]
```diff
@@ -0,0 +1,125 @@
+---
+layout: "ibm"
+page_title: "IBM : ibm_iam_access_group_template"
+description: |-
+  Get information about iam_access_group_template
+subcategory: "IAM Access Groups"
```
Comment (hkantare) [inline website/docs/d/iam_access_group_template.html.markdown:6]: Already we have subgroup  Identity & Access Management (IAM) for other access group res/ds can we use the same

File: website/docs/d/iam_access_group_template_versions.html.markdown [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+# ibm_iam_access_group_template_version
+
+Provides a read-only data source to retrieve information about an ibm_iam_access_group_template_version. You can then reference the fields of the data source in other resources within the same configuration by using interpolation syntax.
+
+## Example Usage
+
+```hcl
+data "ibm_iam_access_group_template_versions" "ibm_iam_access_group_template_versions" {
+	template_id = "template_id"
```
Comment (hkantare) [inline website/docs/d/iam_access_group_template_versions.html.markdown:17]: Can we some sample value or some interpolation syntax to read "template_id"

**Comment by hkantare** [comment]
General comment 
can we remove account_id from all resource/ds where it is described as "Required" and fetch the value from logged in user context
```
userDetails, err := meta.(conns.ClientSession).BluemixUserDetails()
	accountID := userDetails.UserAccount
	createAccountSettingsTemplateOptions.SetAccountID(accountID)
	```

File: ibm/service/iamaccessgroup/resource_ibm_iam_access_group_template_assignment.go (hunk 1 above)
Comment (hariarla) [reply ibm/service/iamaccessgroup/resource_ibm_iam_access_group_template_assignment.go:269]: removed

File: website/docs/d/iam_access_group_template.html.markdown (hunk 3 above)
Comment (hariarla) [reply website/docs/d/iam_access_group_template.html.markdown:6]: Updated

File: website/docs/d/iam_access_group_template_versions.html.markdown (hunk 4 above)
Comment (hariarla) [reply website/docs/d/iam_access_group_template_versions.html.markdown:17]: updated

File: website/docs/r/iam_access_group_template.html.markdown [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  * Constraints: The maximum length is `2` characters. The minimum length is `1` character. The value must match regular expression `/^[0-9]+$/`.
+
+
+## Import
+
+You can import the `ibm_iam_access_group_template` resource by using `id`. The ID of the access group template.
+
+# Syntax
+```
+$ terraform import ibm_iam_access_group_template.iam_access_group_template <id>
```
[the remark this answers is not in the document]
Comment (hariarla) [reply website/docs/r/iam_access_group_template.html.markdown:158]: updated.

File: ibm/service/iamaccessgroup/resource_ibm_iam_access_group_template_assignment.go [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+	waitForAssignment(meta, d, isAccountSettingsTemplateAssigned)
+
+	d.SetId("")
+
+	return nil
+}
+
+func waitForAssignment(meta interface{}, d *schema.ResourceData, refreshFn func(string, interface{}) resource.StateRefreshFunc) (interface{}, error) {
+	log.Printf("Waiting for  (%s) to complete...", d.Id())
+	timeoutStr, isSet := os.LookupEnv(envAGAssignmentTimeoutDurationKey)
```
[the remark this answers is not in the document]
Comment (hariarla) [reply ibm/service/iamaccessgroup/resource_ibm_iam_access_group_template_assignment.go:341]: Done

**Comment by hariarla** [comment]
Screenshot 2023-08-24 at 3 20 45 PM

File: ibm/service/iamaccessgroup/resource_ibm_iam_access_group_template_assignment.go (hunk 2 above)
Comment (hariarla) [reply ibm/service/iamaccessgroup/resource_ibm_iam_access_group_template_assignment.go:299]: As discussed, retaining the same behaviour.

**Comment by hariarla** [comment]
> General comment can we remove account_id from all resource/ds where it is described as "Required" and fetch the value from logged in user context
> 
> ```
> userDetails, err := meta.(conns.ClientSession).BluemixUserDetails()
> 	accountID := userDetails.UserAccount
> 	createAccountSettingsTemplateOptions.SetAccountID(accountID)
> 	```
> 	
> 	We follow this approach in res/ds where account_id is required
> ```

Updated as suggested.

**Comment by hariarla** [comment]
@hkantare implemented your review comments. Review again pls.

**Comment by hariarla** [comment]
Screenshot 2023-08-28 at 11 58 17 AM

**Comment by hkantare** [comment]
account_id needs to be updated in docs

**Comment by hariarla** [comment]
Screenshot 2023-08-29 at 8 00 09 PM

**Review by hkantare — APPROVED** [review]

**Pull request merged**